### PR TITLE
Con 2433 review apprenticeship updates

### PR DESCRIPTION
**Pull request by hkf-tech**

### CODE REVIEW

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingEditApprenticeshipDetails.cs [hunk 1]
```diff
@@ -40,14 +42,17 @@ public async Task VerifyMapperIsCalled()
...
+            _controller.TempData = new TempDataDictionary( Mock.Of<HttpContext>(), Mock.Of<ITempDataProvider>());
+        }
 
         public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestViewModel viewModel)
         {
             return await _controller.EditApprenticeship(viewModel);
         }     
 
-        public void VerifyValidationApiIsCalled()
+        public void VerifyViewModelMapperIsCalled()
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingEditApprenticeshipDetails.cs:55]: we have changed the name of this method but it is still verifying the api call

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using SFA.DAS.CommitmentsV2.Types;
+using static SFA.DAS.CommitmentsV2.Api.Types.Responses.GetPriceEpisodesResponse;
+using SFA.DAS.Encoding;
+
+namespace SFA.DAS.EmployerCommitmentsV2.Web.UnitTests.Mappers.Apprentice
+{
+    public class ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:18]: class name typo

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    {
+        ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTestsFixture fixture;
+
+        [SetUp]
+        public void Setup()
+        {
+            fixture = new ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTestsFixture();
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:25]: Fixture class name typo also

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        public async Task<ConfirmEditApprenticeshipViewModel> Map()
+        {
+          resultViewModl =  await _mapper.Map(source);
+            return resultViewModl;
+        }
+
+        internal void VerifyCommitmentApiIsCalled()
+        {
+            _mockCommitmentsApiClient.Verify(c => c.GetApprenticeship(It.IsAny<long>(), It.IsAny<CancellationToken>()), Times.Once());
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:240]: could verify correct apprenticeshipId is passed to the mocked call

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        }
+
+        internal void VerifyPriceEpisodeIsCalled()
+        {
+            _mockCommitmentsApiClient.Verify(c => c.GetPriceEpisodes(It.IsAny<long>(), It.IsAny<CancellationToken>()), Times.Once());
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:245]: could verify correct apprenticeshipId is passed to the mocked call

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapperTests.cs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                    } }
+                };
+
+                CommitmentApiClient.Setup(x => x.GetApprenticeship(ApprenticeshipId, It.IsAny<CancellationToken>())).Returns(() => Task.FromResult(GetApprenticeshipResponse));
+                CommitmentApiClient.Setup(x => x.GetApprenticeshipUpdates(ApprenticeshipId, It.IsAny<GetApprenticeshipUpdatesRequest>(), It.IsAny<CancellationToken>())).Returns(() => Task.FromResult(GetApprenticeshipUpdatesResponses));
+                CommitmentApiClient.Setup(x => x.GetPriceEpisodes(ApprenticeshipId, It.IsAny<CancellationToken>())).Returns(() => Task.FromResult(priceEpisode));
+                CommitmentApiClient.Setup(x => x.GetTrainingProgramme(It.IsAny<string>(), It.IsAny<CancellationToken>())).Returns(() => Task.FromResult(GetTrainingProgrammeResponse));
+
+
+                Mapper = new ReviewApprenticehipUpdatesRequestToViewModelMapper(CommitmentApiClient.Object);
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapperTests.cs:225]: typo `Apprenticehip`

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapperTests.cs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+        [Test]
+        public async Task OriginalApprenticeship_TrainingName_IsMapped()
+        {
+            var viewModel = await fixture.Map();
+
+            Assert.AreEqual(fixture.ApprenticeshipUpdate.TrainingName, viewModel.TrainingName);
+        }
+
+        public class ReviewApprenticehipUpdatesRequestToViewModelMapperTestsFixture
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapperTests.cs:183]: typo `Apprenticehip`

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 8]
```diff
@@ -544,15 +547,78 @@ public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestVie
...
+            var viewModel = await _modelMapper.Map<ConfirmEditApprenticeshipViewModel>(editApprenticeshipRequestViewModel);
+
+            return View(viewModel);
+        }
+
+        [HttpPost]
+        [Route("{apprenticeshipHashedId}/edit/confirm")]
+        public async Task<IActionResult> ConfirmEditApprenticeship(ConfirmEditApprenticeshipViewModel viewModel)
+        {
+            if (viewModel.ConfirmChanges.Value)
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:569]: why is this Nullable<bool>? If this is null then this controller action will throw an exception

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 9]
```diff
@@ -544,15 +547,78 @@ public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestVie
...
+            var viewModel = await _modelMapper.Map<ReviewApprenticeshipUpdatesRequestViewModel>(request);
+
+            return View(viewModel);
+        }
+
+        [HttpPost]
+        [Route("{apprenticeshipHashedId}/changes/review")]
+        public async Task<IActionResult> ReviewApprenticeshipUpdates(ReviewApprenticeshipUpdatesRequestViewModel viewModel)
+        {
+            if (viewModel.ApproveChanges.Value)
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:600]: why is this Nullable<bool>? If this is null then this controller action will throw an exception

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 10]
```diff
@@ -544,15 +547,78 @@ public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestVie
...
+                var request = new RejectApprenticeshipUpdatesRequest
+                {
+                    ApprenticeshipId = viewModel.ApprenticeshipId,
+                    AccountId = viewModel.AccountId
+                };
+
+                await _commitmentsApiClient.RejectApprenticeshipUpdates(viewModel.ApprenticeshipId, request);
+            }
+
+            return await Task.FromResult(RedirectToAction(nameof(ApprenticeshipDetails), new { viewModel.AccountHashedId, viewModel.ApprenticeshipHashedId }));
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:621]: `await Task.FromResult()` not necessary

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+using SFA.DAS.CommitmentsV2.Api.Types.Responses;
+using SFA.DAS.CommitmentsV2.Shared.Interfaces;
+using SFA.DAS.EmployerCommitmentsV2.Web.Extensions;
+using SFA.DAS.EmployerCommitmentsV2.Web.Models.Apprentice;
+using SFA.DAS.Encoding;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EmployerCommitmentsV2.Web.Mappers.Apprentice
+{
+    public class ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper : IMapper<EditApprenticeshipRequestViewModel, ConfirmEditApprenticeshipViewModel>
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:11]: typo

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            _commitmentApi = commitmentsApi;
+            _encodingService = encodingService;
+        }
+
+        public async Task<ConfirmEditApprenticeshipViewModel> Map(EditApprenticeshipRequestViewModel source)
+        {
+            source.ApprenticeshipId = _encodingService.Decode(source.HashedApprenticeshipId, EncodingType.ApprenticeshipId);
+            source.AccountId = _encodingService.Decode(source.AccountHashedId, EncodingType.AccountId);
+
+            var apprenticehsipTask = _commitmentApi.GetApprenticeship(source.ApprenticeshipId);
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:27]: typo

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            vm.OriginalApprenticeship.EndMonth = apprenticeship.EndDate.Month;
+            vm.OriginalApprenticeship.EndYear = apprenticeship.EndDate.Year;
+
+            if (source.CourseCode != apprenticeship.CourseCode)
+            {
+                vm.CourseCode = source.CourseCode;
+                vm.TrainingName = courseDetails?.TrainingProgramme.Name;
+            }
+            vm.OriginalApprenticeship.CourseCode = apprenticeship.CourseCode;
+            vm.OriginalApprenticeship.TrainingName = courseDetailsOriginal?.TrainingProgramme.Name;
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:107]: could this not be apprenticeship.CourseName?

If this change is made, then the api call above on line 36 could be moved within the if statement.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        private readonly ICommitmentsApiClient _commitmentsApiClient;
+
+        public ReviewApprenticehipUpdatesRequestToViewModelMapper(ICommitmentsApiClient commitmentsApiClient)
+        {
+            _commitmentsApiClient = commitmentsApiClient;
+        }
+
+        public async Task<ReviewApprenticeshipUpdatesRequestViewModel> Map(ReviewApprenticehipUpdatesRequest source)
+        {
+            var updates = await _commitmentsApiClient.GetApprenticeshipUpdates(source.ApprenticeshipId,
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs:23]: these can be done in parallel

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+                if (update.Cost.HasValue)
+                {
+                    var priceEpisodes = await _commitmentsApiClient.GetPriceEpisodes(source.ApprenticeshipId);
+                    vm.OriginalApprenticeship.Cost = priceEpisodes.PriceEpisodes.GetPrice();
+                }
+
+                if (!string.IsNullOrWhiteSpace(update.TrainingCode))
+                {
+                    var courseDetailsOriginal = await _commitmentsApiClient.GetTrainingProgramme(apprenticeship.CourseCode);
+                    vm.OriginalApprenticeship.TrainingName = courseDetailsOriginal?.TrainingProgramme.Name;
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs:66]: could this no be `apprenticeship.CourseName`?

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/Apprentice/EditApprenticeship.cshtml [hunk 16]
```diff
@@ -118,7 +118,7 @@
                     <label asp-for="CourseCode" class="govuk-label govuk-!-font-weight-bold">Apprenticeship training course</label>
                     <span id="course-name-hint" class="govuk-hint">Start typing the name of the course or choose an option from the list </span>
                     @Html.ValidationMessageFor(m => m.CourseCode, null, new { @class = "govuk-error-message", id = "error-message-" + Html.IdFor(m => m.CourseCode) })
-                    <select class="govuk-select govuk-!-width-three-quarters das-autocomplete" asp-for="CourseCode" asp-items="@(new SelectList(Model.Courses, "CourseCode", "Name"))" aria-describedby="course-name-hint" id="trainingCourse">
+                    <select class="govuk-select govuk-!-width-three-quarters" asp-for="CourseCode" asp-items="@(new SelectList(Model.Courses, "CourseCode", "Name"))" aria-describedby="course-name-hint" id="trainingCourse">
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Views/Apprentice/EditApprenticeship.cshtml:121]: did we want to remove the autocomplete?

**Review by cofaulco — COMMENTED** [review]
minor comments

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingEditApprenticeshipDetails.cs (hunk 1 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Controllers/ApprenticeControllerTests/WhenPostingEditApprenticeshipDetails.cs:55]: Renamed it back to Api call, as there is a test for mapper already.

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs (hunk 2 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:18]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs (hunk 3 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:25]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs (hunk 4 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:240]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs (hunk 5 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapperTests.cs:245]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs (hunk 10 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:621]: copy & paste :)

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs (hunk 11 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:11]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs (hunk 12 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:27]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs (hunk 13 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ConfirmEditApprenticehsipRequestToConfirmEditViewModelMapper.cs:107]: made changes for this

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs (hunk 14 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs:23]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs (hunk 15 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ReviewApprenticehipUpdatesRequestToViewModelMapper.cs:66]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/Apprentice/EditApprenticeship.cshtml (hunk 16 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Views/Apprentice/EditApprenticeship.cshtml:121]: it aligns it with what we have on draft apprenticeship.

**Review by cofaulco — APPROVED** [review]

**Pull request merged**